lcmark.1
.TH "lcmark" "1" "January 1, 2016" "" ""
.SH
NAME
.PP
lcmark \- flexible command line CommonMark parser and renderer
.SH
SYNOPSIS
.PP
lcmark [options] [file..]
.SH
DESCRIPTION
.PP
\f[C]lcmark\f[] does what the \f[C]cmark\f[] (https://github.com/jgm/cmark)
program does, with the following enhancements:
.IP \[bu] 2
Support for \f[B]YAML metadata\f[] at the top of the document.
.IP \[bu] 2
Support for \f[B]filters\f[], which allow the document to be transformed between
parsing and rendering, making a large number of customizations possible.
.IP \[bu] 2
Support for \f[B]templates\f[], which allow the body and metadata values to be
embedded into a pre\-defined structure.
.SH
OPTIONS
.PP
\f[C]\-\-to,\-t\f[] \f[I]format\f[]
.PP
Specify format for output.
\f[I]format\f[] can be \f[C]commonmark\f[], \f[C]html\f[], \f[C]man\f[], \f[C]xml\f[], or \f[C]latex\f[].
.PP
\f[C]\-\-output,\-o\f[] \f[I]file\f[]
.PP
Write output to \f[I]file\f[].
.PP
\f[C]\-\-columns,\-c\f[] \f[I]NUMBER\f[]
.PP
Specify number of columns for text wrapping in supported
formats.  The default is 0 = no wrapping.
.PP
\f[C]\-\-filter,\-F\f[] \f[I]file[,file]\f[]
.PP
Filter the parsed AST using a Lua script.  See FILTERS
below for details.
.PP
\f[C]\-\-template,\-T\f[] \f[I]file\f[]
.PP
Insert converted text and metadata into a template.  See TEMPLATES,
below, for template format.
.PP
\f[C]\-\-smart\f[]
.PP
Enable smart typography (straight quotes are turned into
curly quotes, \f[C]\-\-\f[] into en dashes, \f[C]\-\-\-\f[] into em dashes,
\f[C]...\f[] into ellipses).
.PP
\f[C]\-\-safe\f[]
.PP
Suppress raw HTML and unsafe links (\f[C]javascript:\f[], \f[C]file:\f[],
\f[C]vbscript:\f[], and \f[C]data:\f[], except for \f[C]image/png\f[], \f[C]image/gif\f[],
\f[C]image/jpeg\f[], or \f[C]image/webp\f[] mime types).
.PP
\f[C]\-\-hardbreaks\f[]
.PP
Render softbreak elements (newlines in paragraphs) as hard
line breaks.
.PP
\f[C]\-\-sourcepos\f[]
.PP
Include source position information in html/xml attributes.
.PP
\f[C]\-\-version,\-V\f[]
.PP
Print version information.
.PP
\f[C]\-\-help,\-h\f[]
.PP
This message
.SH
METADATA
.PP
The YAML metadata section (if present) must occur at the beginning of the
document.  It begins with a line containing \f[C]\-\-\-\f[] and ends with a line
containing \f[C]...\f[] or \f[C]\-\-\-\f[].  Between these, a YAML key/value map is expected.
.PP
String values found in the metadata will be parsed and rendered as
CommonMark. If a string value contains only a single paragraph, it will be
rendered as an inline string.
.SH
TEMPLATES
.PP
By default, \f[C]lcmark\f[] will produce a fragment.  If the \f[C]\-\-template\f[]
option is specified, it will insert this fragment into a
template, producing a standalone document with appropriate
header and footer.  The template is sought first in the working
directory, then in \f[C]templates\f[], and finally in
\f[C]$HOME/lcmark/templates\f[]. If no extension is given, the name of
the writer will be used as an extension. So, for example, one
can put the template \f[C]letter.html\f[] in the
\f[C]$HOME/lcmark/templates\f[] directory, and use it anywhere with
\f[C]lcmark \-\-template letter\f[].
.PP
Variables are taken from YAML metadata; string fields are interpreted
as CommonMark and rendered appropriately for the output format.
The following additional variables are set automatically:
.IP \[bu] 2
\f[C]body\f[]: the document body
.PP
\f[C]lcmark\f[] supports a small subset of the templating language used by
pandoc (http://pandoc.org), and \f[C]lcmark\f[] templates can be used with pandoc
(with the caveat that pandoc sets many variables automatically that \f[C]lcmark\f[]
does not).
.PP
\f[C]nil\f[], \f[C]false\f[] and empty tables are considered to be "falsy" values.
Any other value is considered to be "truthy".
.PP
A quick guide:
.IP \[bu] 2
The only special character in templates is \f[C]$\f[].  To get
a literal \f[C]$\f[] character, use \f[C]$$\f[].
.IP \[bu] 2
\f[C]$name$\f[] will be replaced with the value of the \f[C]name\f[]
metadata field.  Variable names can contain alphanumerics,
\f[C]\-\f[], and \f[C]_\f[].
.IP \[bu] 2
\f[C]$name.subname$\f[] will be replaced with the value of the
\f[C]subname\f[] field of the \f[C]name\f[] metadata field (assumed to
be a map).  More indexes can be changed together this way.
.IP \[bu] 2
\f[C]$if(name)$...$endif$\f[] will be replaced by the content
in \f[C]...\f[] if the value of the \f[C]name\f[] metadata field is
"truthy", otherwise by nothing.  \f[C]...\f[] may contain
nested templating directives.
.IP \[bu] 2
\f[C]$if(name)$...$else$,,,$endif$\f[] will be
replaced by the content in \f[C]...\f[] if \f[C]name\f[] has a truthy
value, and by the content in \f[C],,,\f[] otherwise.  Both
\f[C]...\f[] and \f[C],,,\f[] may contain nested templating directives.
.IP \[bu] 2
\f[C]$for(name)$...$endfor$\f[] is a loop, producing
successive concatenated copies of \f[C]...\f[]. If the value
of \f[C]name\f[] is a non\-empty table, then in each occurrence
of \f[C]...\f[], the value of \f[C]name\f[] will be replaced by a
different element from the table (in order).  For example,
\f[C]$for(authors)$$authors$$endfor$\f[] will concatenate
all the values of the \f[C]authors\f[] table.
.PP
Otherwise, if the value of \f[C]name\f[] isn't a table, the loop
behaves like an \f[C]if\f[].
.IP \[bu] 2
\f[C]$for(name)$...$sep$,,,$endfor$\f[] behaves like the above,
except that the content in \f[C],,,\f[] is inserted between each
copy of \f[C]...\f[].  \f[C],,,\f[] supports nested templating directives.
.PP
Additionally, if newlines occurs directly after \f[B]both\f[] \f[C]$for()$\f[] and
\f[C]$endfor$\f[] (or \f[C]$if()$\f[] and \f[C]$endif$\f[]), they will be ignored.  This is to
prevent spurious blank lines in the rendered document if the template contains
many directives that span multiple lines and evaluate to false.
.PP
For examples, see the \f[C]templates/\f[] directory in the source
repository.
.SH
FILTERS
.PP
Filters modify the parsed document prior to rendering.
.PP
A filter is a function that takes three arguments (\f[C]doc\f[], \f[C]meta\f[], \f[C]to\f[]), where
\f[C]doc\f[] is a cmark node, \f[C]meta\f[] is the YAML metadata as a (potentially nested) Lua
table with all strings replaced with cmark nodes, and \f[C]to\f[] is a string
specifying the output format.  The filter may destructively modify \f[C]doc\f[] and
\f[C]meta\f[].
.PP
When loading filters, \f[C]lcmark\f[] automatically populates the filter function's
environment with the functions and values provided by
\f[C]cmark\-lua\f[] (https://github.com/jgm/cmark\-lua) so that any \f[C]cmark\f[] functions do
not have to be qualified with \f[C]cmark.\f[].
.PP
For examples, see the \f[C]filters/\f[] directory in the source
repository.
.PP
The arguments to \f[C]\-\-filter\f[] should be Lua scripts that \f[C]return\f[]
a filter function, as defined above.  Filters will be run in the
order listed.  Filters are applied to the root document node,
not to metadata (although a filter can operate on metadata if
desired).
.SH
EXAMPLES
.IP
.nf
\f[C]
lcmark
\f[]
.fi
.PP
acts as a filter, reading markdown from stdin and writing
HTML to stdout.
.IP
.nf
\f[C]
lcmark \-\-smart \-t latex
\f[]
.fi
.PP
acts as a filter, reading markdown with smart typography
and definition list extensions from stdin, and writing
LaTeX to stdout.
.IP
.nf
\f[C]
lcmark \-t latex \-o mybook.tex ch{1,2,3}.txt references.txt
\f[]
.fi
.PP
reads \f[C]ch1.txt\f[], \f[C]ch2.txt\f[], \f[C]ch3.txt\f[], and \f[C]references.txt\f[],
concatenates them, and converts the result from markdown to LaTeX.
.IP
.nf
\f[C]
lcmark \-\-template letter \-t latex \-o myletter.tex myletter.txt
\f[]
.fi
.PP
produces a LaTeX file using the template \f[C]letter.latex\f[].

.SH AUTHORS
John MacFarlane


lcmark.lua
local cmark = require("cmark")
local lpeg = require("lpeg")

local S, C, P, R, V, Ct =
  lpeg.S, lpeg.C, lpeg.P, lpeg.R, lpeg.V, lpeg.Ct
local nl = P"\r\n" + P"\r" + P"\n"
local sp = S" \t"^0

local lcmark = {}

lcmark.version = "0.29.0"

lcmark.writers = {
  html = function(d, opts, _) return cmark.render_html(d, opts) end,
  man = cmark.render_man,
  xml = function(d, opts, _) return cmark.render_xml(d, opts) end,
  latex = cmark.render_latex,
  commonmark = cmark.render_commonmark
}

local default_yaml_parser = nil

local function try_load(module_name, func_name)
  if default_yaml_parser then -- already loaded; skip
    return
  end

  local success, loaded = pcall(require, module_name)

  if not success then
    return
  end
  if type(loaded) ~= "table" or type(loaded[func_name]) ~= "function" then
    return
  end

  default_yaml_parser = loaded[func_name]
  lcmark.yaml_parser_name = module_name .. "." .. func_name
end

try_load("lyaml", "load")
try_load("yaml", "load") -- must come before yaml.eval
try_load("yaml", "eval")

-- the reason yaml.load must come before yaml.eval is that the 'yaml' library
-- prints error messages if you try to index non-existent fields such as 'eval'


local toOptions = function(opts)
  if type(opts) == 'table' then
    return (cmark.OPT_VALIDATE_UTF8 + cmark.OPT_NORMALIZE +
      (opts.smart and cmark.OPT_SMART or 0) +
      (opts.safe and 0 or cmark.OPT_UNSAFE) +
      (opts.hardbreaks and cmark.OPT_HARDBREAKS or 0) +
      (opts.sourcepos and cmark.OPT_SOURCEPOS or 0)
      )
  else
     return opts
  end
end

-- walk nodes of table, applying a callback to each
local function walk_table(table, callback, inplace)
  assert(type(table) == 'table')
  local new = {}
  local res
  for k, v in pairs(table) do
    if type(v) == 'table' then
      res = walk_table(v, callback, inplace)
    else
      res = callback(v)
    end
    if not inplace then
      new[k] = res
    end
  end
  if not inplace then
    return new
  end
end

-- We inject cmark into environment where filters are
-- run, so users don't need to qualify each function with 'cmark.'.
local defaultEnv = setmetatable({}, { __index = _G })
for k,v in pairs(cmark) do
  defaultEnv[k] = v
end

local loadfile_with_env
if setfenv then
  -- Lua 5.1/LuaJIT
  loadfile_with_env = function(filename)
    local result, msg = loadfile(filename)
    if result then
      return setfenv(result, defaultEnv)
    else
      return result, msg
    end
  end
else
  -- Lua 5.2+
  loadfile_with_env = function(filename)
    return loadfile(filename, 't', defaultEnv)
  end
end

-- Loads a filter from a Lua file and populates the loaded function's
-- environment with all the fields from `cmark-lua`.
-- Returns the filter function on success, or `nil, msg` on failure.
function lcmark.load_filter(filename)
  local result, msg = loadfile_with_env(filename)
  if result then
    local evaluated = result()
    if type(evaluated) == 'function' then
        return evaluated
    else
        return nil, ("Filter " .. filename .. " returns a " ..
                    type(evaluated) .. ", not a function")
    end
  else
    return nil, msg
  end
end

-- Render a metadata node in the target format.
local render_metadata = function(node, writer, options, columns)
  local firstblock = cmark.node_first_child(node)
  if cmark.node_get_type(firstblock) == cmark.NODE_PARAGRAPH and
     not cmark.node_next(firstblock) then
     -- render as inlines
     local ils = cmark.node_new(cmark.NODE_CUSTOM_INLINE)
     local b = cmark.node_first_child(firstblock)
     while b do
        local nextb = cmark.node_next(b)
        cmark.node_append_child(ils, b)
        b = nextb
     end
     local result = string.gsub(writer(ils, options, columns), "%s*$", "")
     cmark.node_free(ils)
     return result
  else -- render as blocks
     return writer(node, options, columns)
  end
end

-- Iterate over the metadata, converting to cmark nodes.
-- Returns a new table.
local convert_metadata = function(table, options)
  return walk_table(table,
                    function(s)
                      if type(s) == "string" then
                        return cmark.parse_string(s, options)
                      elseif type(s) == "userdata" then
                        return tostring(s)
                      else
                        return s
                      end
                    end, false)
end

local yaml_begin_line = P"---" * sp * nl
local yaml_end_line = (P"---" + P"...") * sp * nl
local yaml_content_line = -yaml_end_line * P(1 - S"\r\n")^0 * nl
local yaml_block = yaml_begin_line * (yaml_content_line^1 + sp) * yaml_end_line

-- Parses document with optional front YAML metadata; returns document,
-- metadata.
local parse_document_with_metadata = function(inp, parser, options)
  local metadata = {}
  local meta_end = lpeg.match(yaml_block, inp)
  if meta_end then
    if meta_end then
      local ok, yaml_meta, err = pcall(parser, string.sub(inp, 1, meta_end))
      if not ok then
        return nil, yaml_meta -- the error message
      elseif not yaml_meta then -- parser may return nil, err instead of error
        return nil, tostring(err)
      end
      if type(yaml_meta) == 'table' then
        metadata = convert_metadata(yaml_meta, options)
        if type(metadata) ~= 'table' then
          metadata = {}
        end
        -- We insert blank lines where the header was, so sourcepos is accurate:
        inp = string.gsub(string.sub(inp, 1, meta_end), '[^\n\r]+', '') ..
           string.sub(inp, meta_end)
      end
    end
  end
  local doc = cmark.parse_string(inp, options)
  return doc, metadata
end

-- Apply a compiled template to a context (a dictionary-like
-- table).
function lcmark.apply_template(m, ctx)
  if type(m) == 'function' then
    return m(ctx)
  elseif type(m) == 'table' then
    local buffer = {}
    for i,v in ipairs(m) do
      buffer[i] = lcmark.apply_template(v, ctx)
    end
    return table.concat(buffer)
  else
    return tostring(m)
  end
end

local get_value = function(var, ctx)
  local result = ctx
  assert(type(var) == 'table')
  for _,varpart in ipairs(var) do
    if type(result) ~= 'table' then
      return nil
    end
    result = result[varpart]
    if result == nil then
      return nil
    end
  end
  return result
end

local set_value = function(var, newval, ctx)
  local result = ctx
  assert(type(var) == 'table')
  for i,varpart in ipairs(var) do
    if i == #var then
      -- last one
      result[varpart] = newval
    else
      result = result[varpart]
      if result == nil then
        return nil
      end
    end
  end
  return true
end

local is_truthy = function(val)
  local is_empty_tbl = type(val) == "table" and #val == 0
  return val and not is_empty_tbl
end

-- if s starts with newline, remove initial and final newline
local trim = function(s)
  if s:match("^[\r\n]") then
    return s:gsub("^[\r]?[\n]?", ""):gsub("[\r]?[\n]?$", "")
  else
    return s
  end
end

local conditional = function(var, ifpart, elsepart)
  return function(ctx)
    local result
    if is_truthy(get_value(var, ctx)) then
      result = lcmark.apply_template(ifpart, ctx)
    elseif elsepart then
      result = lcmark.apply_template(elsepart, ctx)
    else
      result = ""
    end
    return trim(result)
  end
end

local forloop = function(var, inner, sep)
  return function(ctx)
    local val = get_value(var, ctx)
    local vs
    if not is_truthy(val) then
      return ""
    end
    if type(val) == 'table' then
      vs = val
    else
      -- if not a table, just iterate once
      vs = {val}
    end
    local buffer = {}
    for i,v in ipairs(vs) do
      set_value(var, v, ctx) -- set temporary context
      buffer[#buffer + 1] = lcmark.apply_template(inner, ctx)
      if sep and i < #vs then
        buffer[#buffer + 1] = lcmark.apply_template(sep, ctx)
      end
      set_value(var, val, ctx) -- restore original context
    end
    local result = lcmark.apply_template(buffer, ctx)
    return trim(result)
  end
end

-- Template syntax.
local TemplateGrammar = Ct{"Main",
  Main = V"Template" * (-1 + lpeg.Cp()),
  Template = Ct((V"Text" +
                 V"EscapedDollar" +
                 V"ConditionalNl" +
                 V"Conditional" +
                 V"ForLoopNl" +
                 V"ForLoop" +
                 V"Var")^0),
  EscapedDollar = P"$$" / "$",
  -- the Nl forms eat an extra newline after the end, if the
  -- opening if() or for() ends with a newline.  This is to avoid
  -- excess blank space when a document contains many ifs or fors
  -- that evaluate to false.
  ConditionalNl = P"$if(" * Ct(V"Variable") * P")$" * nl * Ct(V"Template") *
    (P"$else$" * Ct(V"Template"))^-1 * P"$endif$" * nl / conditional,
  ForLoopNl = P"$for(" * Ct(V"Variable") * P")$" * nl * Ct(V"Template") *
    (P"$sep$" * Ct(V"Template"))^-1 * P"$endfor$" * nl / forloop,
  Conditional = P"$if(" * Ct(V"Variable") * P")$" * Ct(V"Template") *
    (P"$else$" * Ct(V"Template"))^-1 * P"$endif$" / conditional,
  ForLoop = P"$for(" * Ct(V"Variable") * P")$" * Ct(V"Template") *
    (P"$sep$" * Ct(V"Template"))^-1 * P"$endfor$" / forloop,
  Text = C((1 - P"$")^1),
  Reserved = P"if$" + P"endif$" + P"else$" + P"for$" + P"endfor$" + P"sep$",
  VarPart = (R"az" + R"AZ" + R"09" + S"_-")^1,
  Variable = C(V"VarPart") * (P"." * C(V"VarPart"))^0,
  Var = P"$" * - V"Reserved" * Ct(V"Variable") * P"$" /
    function(var)
      return function(ctx)
        local val = get_value(var, ctx)
        if is_truthy(val) then
          return tostring(val)
        else
          return ""
        end
      end
    end,
}

-- Compiles a template string into an  arbitrary template object
-- which can then be passed to `lcmark.apply_template()`.
-- Returns the template object on success, or `nil, msg` on failure.
lcmark.compile_template = function(tpl)
  local matches = lpeg.match(TemplateGrammar, tpl, nil)
  if matches[2] == nil then
    if matches[1] == nil then
      return nil, "parse failed at the end of the template"
    else
      return matches[1]
    end
  else
    local line_num = 1
    local parse_failure_pos = matches[2]
    tpl:sub(1,parse_failure_pos):gsub('[^\n]*[\n]',
            function() line_num = line_num + 1 end)
    return nil, ("parse failure at line " .. line_num ..
                  ": '" .. string.sub(tpl, parse_failure_pos,
                                      parse_failure_pos + 40) .. "'")
  end
end

-- Compiles and applies a template string to a context table.
-- Returns the  resulting document string on success, or
-- `nil, msg` on failure.
function lcmark.render_template(tpl, ctx)
  local compiled_template, msg = lcmark.compile_template(tpl)
  if not compiled_template then
    return nil, msg
  end
  return lcmark.apply_template(compiled_template, ctx)
end

-- Converts `inp` (a CommonMark formatted string) to the output
-- format specified by `to` (a string; one of `html`, `commonmark`,
-- `latex`, `man`, or `xml`).  `options` is a table with the
-- following fields (all optional):
-- * `smart` - enable "smart punctuation"
-- * `hardbreaks` - treat newlines as hard breaks
-- * `safe` - filter out potentially unsafe HTML and links
-- * `sourcepos` - include source position in HTML and XML output
-- * `filters` - an array of filters to run (see `load_filter` above)
-- * `columns` - column width, or 0 to preserve wrapping in input
-- * `yaml_metadata` - whether to parse initial YAML metadata block
-- * `yaml_parser` - a function to parse YAML with (see
--    [YAML Metadata](#yaml-metadata))
-- Returns `body`, `meta` on success, where `body` is the rendered
-- document body and `meta` is the YAML metadata as a table. If the
-- `yaml_metadata` option is false or if the document contains no
-- YAML metadata, `meta` will be an empty table. In case of an
-- error, the function returns `nil, nil, msg`.
function lcmark.convert(inp, to, options)
  local writer = lcmark.writers[to]
  if not writer then
    return nil, nil, ("Unknown output format " .. tostring(to))
  end
  local opts, columns, filters, yaml_metadata, yaml_parser
  if options then
     opts = toOptions(options)
     columns = options.columns or 0
     filters = options.filters or {}
     yaml_metadata = options.yaml_metadata
     yaml_parser = options.yaml_parser or default_yaml_parser
  else
     opts = cmark.OPT_DEFAULT
     columns = 0
     filters = {}
     yaml_metadata = false
     yaml_parser = default_yaml_parser
  end
  if not yaml_parser then
    error("no YAML libraries were found and no yaml_parser was specified")
  end
  local doc, meta
  if yaml_metadata then
    doc, meta = parse_document_with_metadata(inp, yaml_parser, opts)
    if not doc then
      return nil, nil, ("YAML parsing error: " .. meta)
    end
  else
    doc = cmark.parse_string(inp, opts)
    meta = {}
  end
  if not doc then
    return nil, nil, "Unable to parse document"
  end
  for _, f in ipairs(filters) do
    -- do we want filters to apply automatically to metadata?
    -- better to let users do this manually when they want to.
    -- walk_table(meta, function(node) f(node, meta, to) end, true)
    local ok, msg = pcall(function() f(doc, meta, to) end)
    if not ok then
      return nil, nil, ("Error running filter:\n" .. msg)
    end
  end
  local body = writer(doc, opts, columns)
  local data = walk_table(meta,
                          function(node)
                            if type(node) == "userdata" then
                              return render_metadata(node, writer, opts, columns)
                            else
                              return node
                            end
                          end, false)
  -- free memory allocated by libcmark
  cmark.node_free(doc)
  walk_table(meta,
             function(node)
               if type(node) == "userdata" then
                 cmark.node_free(node)
               end
             end, true)
  return body, data
end

return lcmark
